TinyProfile parser that saves profile data in a machine readible format

EtherealMC-Bit opened the pull request:
This PR adds a script that parses the saved stdout information from a WarpX run and parses the TinyProfile output at the end of it. This output is then saved as a JSON file.

EtherealMC-Bit committed:
Start to the profile parser

EtherealMC-Bit committed:
Parsed profile data from inclusive data

EtherealMC-Bit committed:
Finished functionality for writing profiling data

PhilMiller commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,136 @@
+            line = re.sub("\s+", " ", line)
+            values = line.split(" ")
+
+            name = values[0]
```
Warp had some profiling strings that contained spaces in their names. I don't know if WarpX is better/stricter about this.

If it's not, you could probably make this all work by working backwards from the end of the split line, with values coming as

```python
max_percent = foo(values[-1])
excl_max = float(values[-2])
...
name = values[0:-5].join(" ") # check that -5 for correctness
```

EtherealMC-Bit replied:
Even if WarpX is stricter, it's a good idea to be on the safe side.

PhilMiller commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,136 @@
+
+    full_profile = FullProfile(profile_lines)
+
+    full_profile.parse_and_write_full_profiling_output()
```
For the sake of good design, and potential programmatic use that won't go through reading the JSON file in, please just call `full_profile.parse(); full_profile.write_file()` here, and don't have a method `do_this_and_that`

PhilMiller replied:
One motivation is that Hatchet can be used in scripts, consuming the dictionary directly, rather than having to read in the JSON

EtherealMC-Bit replied:
> For the sake of good design, and potential programmatic use that won't go through reading the JSON file in, please just call full_profile.parse(); full_profile.write_file() here, and don't have a method do_this_and_that

Yep that's much better, thanks.

PhilMiller commented:
With confirmation that Hatchet can read in the output, I'll be +1 on this once the couple comments I've made are addressed.

EtherealMC-Bit committed:
Capture profile names with spaces, and better design recommended from code review

EtherealMC-Bit committed:
Added message for the number of profiling strings parsed

EtherealMC-Bit committed:
Changed the formatting of the JSON file so that it can be read by Hatchet

EtherealMC-Bit committed:
Reworded comment that referenced deleted code

PhilMiller commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,150 @@
+            excl_avg = float(values[-3])
+            excl_min = float(values[-4])
+            n_calls = int(values[-5])
+            name = "".join(values[0:-5])
```
Could you use a space or underscore for the joining character, rather than an empty string, so that the resulting analysis will have readable output?

PhilMiller replied:
Ditto below

EtherealMC-Bit replied:
Great idea

PhilMiller commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,150 @@
+        del self.lines[0:5]
+        index = 0
+        count = 0
+        for line in self.lines:
```
Could you factor out this whole `for` loop into a separate function and then call it here and in place of its copy just below? I think it would just need to take a string `"incl"` or `"excl"` as an argument to use as a prefix for the metric names.

EtherealMC-Bit replied:
Yep you're right that'll work well

EtherealMC-Bit committed:
For function names with spaces, replace spaces with _ intead of removing them

EtherealMC-Bit committed:
Fixed an issue with inclusive profiles

EtherealMC-Bit committed:
Removed some code duplication by making a function for parsing a section of profiling data

KZhu-ME commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,148 @@
+            raise RuntimeError("write_file called before profile data was parsed!")
+
+        print("Writing profile data json file...")
+        with open("profile_data.json", "w") as data_file:
```
Would it be useful to have the json file be saved in the same directory as the stdout file by default, and if needed, give the ability for the user to specify the save location of the json file in the command line? This way, we don't need to run this script from the directory we want the json to be saved in.

KZhu-ME commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,148 @@
+
+        print("Finished writing json file as 'profile_data.json'.")
+
+def main():
+    if len(sys.argv) != 2:
+        raise RuntimeError(f"Usage: {sys.argv[0]} <path to stdout file>")
```
Can you add the ability to pass the stdout file to the parser through the main() function call as well as through the command line? That way, we can just call this main function at the end of a warpx script, so it can automatically generate the json, instead of having to do it by hand. One way I can think of doing it is to add a path argument to `main()` remove the sys.argv check from `main`. Instead add the check for command line arguments to the `__name__ == "__main__"` part, and pass `sys.argv[1]` to the main function as an argument if it exists.

EtherealMC-Bit replied:
This is a really really great idea, definitely makes using the script easy if I can get it to work

EtherealMC-Bit replied:
@KZhu-ME The problem with this is that even the very end of a python script is before the TinyProfiler output is printed. It would be possible to call that from _libwarpx, but _libwarpx doesn't know about mewarpx stuff.

KZhu-ME replied:
Ooooh yeah I see, that wouldn't work. I still think it's worth moving the sys.argv check to  the `if name == main:` block, since it seems cleaner that way.

roelof-groenewald replied:
If you really wanted to do that you could add this to the bottom of mwxrun.py (you need to import atexit as well)

```
@atexit.register
def exit_handler():
    atexit.unregister(_libwarpx.finalize)
    _libwarpx.finalize()
    print('My application is ending!')
    print('Insert profiling summary here.')
```

EtherealMC-Bit replied:
Wow I've never heard of this. Is this something you'd be interested in? Automatically writing the JSON file seems like it would be nice to me.

roelof-groenewald replied:
Yeah it sounds like a worthwhile change to me as well.

KZhu-ME reviewed (changes requested):
Just a couple of comments. Looks good though!

roelof-groenewald reviewed (commented):
@MKieburtz In principle this could be an made into an upstream PR if you wanted to since it doesn't do anything specific to our library. You certainly don't need to do that, just wanted to point it out in case you feel like contributing it to the overall WarpX project (if so I think it can probably go in `WarpX/Tools/PostProcessing`).

EtherealMC-Bit committed:
Suggestions from Kevin

- Added the option for the user to input the write directory for the JSON file
- Moved the logic for reading in the command line args into the __name__ == "__main__" block

EtherealMC-Bit committed:
If stdout of a simulation was sent to output.txt generate JSON file

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,160 @@
+                self.write_dir = self.stdout_path[0 : self.stdout_path.rindex("/")] + "/"
+            else:
+                self.write_dir = "./"
+        elif "/" not in self.write_dir:
+            self.write_dir += "/"
```
This can be removed if you use `os.path.join` whenever you are concatenating strings into a path.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,160 @@
+    """
+    Class for parsing, storing and writing the full profile data
+    """
+    def __init__(self, lines, stdout_path, write_dir):
```
Maybe making `write_dir` an optional argument (with default `None`) would be better, then just checking below if it is `None` and if so set it to be the same as `self.stdout_path`.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,160 @@
+            raise RuntimeError("write_file called before profile data was parsed!")
+
+        print("Writing profile data json file...")
+        with open(self.write_dir + "profile_data.json", "w") as data_file:
```
Like here for example, using `os.path.join(self.write_dir, "profile_data.json")` avoids you having to worry about whether the write directory ended with a `/`.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,160 @@
+            break
+
+        if reached_profile:
+            profile_lines.append(line.replace("\n", ""))
```
I believe `line.strip()` would do the same thing.

EtherealMC-Bit committed:
Use a space instead of \s for regex

EtherealMC-Bit committed:
Suggested changes from code review

roelof-groenewald commented on mewarpx/mewarpx/mwxrun.py:
```diff
@@ -369,3 +372,11 @@ def calc_Schottky_weight(self, species_name, pre_fac):
+    _libwarpx.finalize()
+
+    if os.path.isfile("output.txt"):
+        profileparser.main("output.txt", "")
```
```suggestion
        profileparser.main("output.txt")
```

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,158 @@
+    full_profile.write_file()
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2 or len(sys.argv) > 3:
+        raise RuntimeError(f"Usage: {sys.argv[0]} path_to_stdout_file [write_directory]")
+
+    path = sys.argv[1]
+    if len(sys.argv) == 3:
+        write_dir = sys.argv[2]
+        main(path, write_dir)
+    else:
+        main(path)
```
This can be improved by using `argparse` (see https://docs.python.org/3/library/argparse.html) in which case you won't need to rely on the order in which arguments are supplied but can instead look for specific arguments (and you can set arguments to be required or optional). This will also allow you to avoid the `if len(sys.argv) == 3:` logic since you can give `write_dir` a default value `None`.

EtherealMC-Bit replied:
Thanks yeah this seems like the place to use that module.

roelof-groenewald commented on mewarpx/mewarpx/utils_store/profileparser.py:
```diff
@@ -0,0 +1,158 @@
+        self.write_dir = write_dir
+
+        if self.write_dir is None:
+            if "/" in self.stdout_path:
+                self.write_dir = self.stdout_path[0 : self.stdout_path.rindex("/")]
+            else:
+                self.write_dir = "./"
```
I believe this can be achieved more cleanly with
```suggestion
            self.write_dir = os.path.dirname(self.stdout_path)
```
but you should confirm that (and I guess you need the extra import).

EtherealMC-Bit committed:
Used argparse to parse arguments when called from the command line

EtherealMC-Bit committed:
Simplify logic for when write_dir is None

EtherealMC-Bit committed:
Remove unneeded blank line

KZhu-ME reviewed (approved):
Looks good!

EtherealMC-Bit committed:
Put all the profiling data in the same frame

EtherealMC-Bit commented:
Alright after one suggestion from Phil, I think this is good to go @roelof-groenewald.

EtherealMC-Bit committed:
Merge memaster into mkieburtz/profiling_parser

EtherealMC-Bit committed:
Forgot to save the file...
